src/hl/attribute.rs

```diff
@@ -0,0 +1,296 @@
+    h5a::{H5A_info_t, H5A_operator2_t, H5Acreate2, H5Aiterate2},
+};
+
+use crate::internal_prelude::*;
+
+/// Represents the HDF5 attribute object.
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct Attribute(Handle);
+
+impl ObjectClass for Attribute {
+    const NAME: &'static str = "attribute";
+    const VALID_TYPES: &'static [H5I_type_t] = &[H5I_ATTR];
+
+    fn from_handle(handle: Handle) -> Self {
+        Self(handle)
+    }
+
+    fn handle(&self) -> &Handle {
+        &self.0
+    }
+
+    // TODO: short_repr()
+}
+
+impl Debug for Attribute {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.debug_fmt(f)
+    }
+}
+
+impl Deref for Attribute {
+    type Target = Container;
+
+    fn deref(&self) -> &Container {
+        unsafe { self.transmute() }
+    }
+}
+
+impl Attribute {
+    /// Returns names of all the members in the group, non-recursively.
+    pub fn attribute_names<T: ObjectClass>(obj: &T) -> Result<Vec<String>> {
+        extern "C" fn attributes_callback(
+            _id: hid_t, attr_name: *const c_char, _info: *const H5A_info_t, op_data: *mut c_void,
+        ) -> herr_t {
+            std::panic::catch_unwind(|| {
+                let other_data: &mut Vec<String> = unsafe { &mut *(op_data as *mut Vec<String>) };
+                other_data.push(string_from_cstr(attr_name));
+                0 // Continue iteration
+            })
+            .unwrap_or(-1)
+        }
+
+        let callback_fn: H5A_operator2_t = Some(attributes_callback);
+        let iteration_position: *mut hsize_t = &mut { 0_u64 };
+        let mut result: Vec<String> = Vec::new();
+        let other_data: *mut c_void = &mut result as *mut _ as *mut c_void;
+
+        h5call!(H5Aiterate2(
+            obj.handle().id(),
+            H5_index_t::H5_INDEX_NAME,
+            H5_iter_order_t::H5_ITER_INC,
+            iteration_position,
+            callback_fn,
+            other_data
+        ))?;
+
+        Ok(result)
+    }
+}
+
+#[derive(Clone)]
+pub struct AttributeBuilder<T> {
+    packed: bool,
+    filters: Filters,
+    parent: Result<Handle>,
+    track_times: bool,
+    phantom: PhantomData<T>,
+}
+
+impl<T: H5Type> AttributeBuilder<T> {
+    /// Create a new dataset builder and bind it to the parent container.
+    pub fn new(parent: &Group) -> Self {
+        h5lock!({
+            // Store the reference to the parent handle and try to increase its reference count.
+            let handle = Handle::try_new(parent.id());
+            if let Ok(ref handle) = handle {
+                handle.incref();
+            }
+
+            Self {
+                packed: false,
+                filters: Filters::default(),
+                parent: handle,
+                track_times: false,
+                phantom: PhantomData,
+            }
+        })
+    }
+
+    /// Create a new dataset builder and bind it to the parent container.
+    pub fn new_from_dataset(parent: &Dataset) -> Self {
```

Comment:
This could be called `new_in_dataset`

Reply:
I know I'm not the owner of the PR, but why? This seems to create a *new* AttributeBuilder *from* or maybe *with* the given *dataset*.

Reply:
@SoftwareApe You are very welcome to leave suggestions! I think we could go for `new` which accepts a `Location` instead, as every named location can have an attribute

Reply:
yep I think a single `new(parent: &Location)` method will suffice. changed accordingly.